Add timetracking endpoints proposal

jenssegers opened the pull request.

jenssegers committed:
Add timetracking endpoints proposal

jenssegers committed:
Rename topic to work type

WouterSioen commented on apiary.apib:
```diff
@@ -1855,6 +1855,272 @@ Update a milestone.
+                + started_at: `2017-04-26T10:01:49+00:00` (string, required)
+                + duration: 3600 (number, required) - In seconds
+        + description (string, optional)
+        + subject (object, optional)
```
Q: shouldn't all these properties be nullable to allow partial updates?

 :ok: by @WouterSioen on 19/03/2018 09:58:04

jenssegers committed:
Make non-required fields nullable in update

WouterSioen reviewed (dismissed):
Let's do this

WouterSioen commented on apiary.apib:
```diff
@@ -2192,6 +2192,272 @@ Update a milestone.
+    + Attributes (object)
+        + data (object)
+            + type: `timetracking` (string)
+            + id: `4ff475b2-f45b-4803-800e-350b9c02aeca` (string)
```
Is it just me or is it strange that we return the same type here for an "ongoing timetracking" as we do for a "submitted timetracking", even though we can't fetch an ongoing timetracking using timetracking.info?

It would be more logical to me to name it ongoingTimeTracking as long as it's not submitted and thus fetchable in timeTracking.info.

jenssegers replied:
On the other hand, of we don't return an id, you are also not able to update information related to a running timer. For me it's ok to change the proposal to return a `204 No Content`, but then we need to be aware of that too.

WouterSioen replied:
So you meant the `timetracking.info` and `imetracking.update` where meant to also work with "ongoing" timetrackings?

I can see the use case in allowing it for `timetracking.update`, because otherwise you can't edit f.e. the description of a running timer.

 :ok: by @WouterSioen on 28/03/2018 10:30:05

WouterSioen commented on apiary.apib:
```diff
@@ -2192,6 +2192,272 @@ Update a milestone.
++ Request (application/json)
+
+    + Attributes (object)
+        + ended_at: `2017-04-26T10:01:49+00:00` (string, optional) - If not provided, current time will be used
```
Shouldn't this endpoint return the `id` of the added timetracking?

This will be different then the `id` returned from `timetracking.start`, because it's another datatype (ongoing vs finished timetracking) + the stopped timetracking can be split here when it spanned multiple days.

 :ok: by @WouterSioen on 28/03/2018 10:30:11

WouterSioen committed:
Rename ongoing time tracking endpoints to timer

Also make a difference in returned types for added items: timer for
"ongoing" and time tracking for "submitted".

WouterSioen committed:
Put the time tracking and timer endpoints together

WouterSioen commented:
Up for re-approve @jenssegers @stivni @mrnnn @JordyDeGroeve @dietrichm 

Note that I didn't add `timer.update` yet in this PR since this can be released later in a separate PR as an extra feature.

JordyDeGroeve reviewed (approved):
LGTM. Renaming to `Timer` makes things clearer.

WouterSioen committed:
Make endpoints use the plural form

WouterSioen committed:
Build previous change

WouterSioen committed:
Use the correct plural of timeTracking

According to the Cambridge English dictionary, "tracking" is a noun that
is not countable and thus has no plural form.

https://dictionary.cambridge.org/dictionary/english/tracking

jenssegers committed:
:pencil2: Split time tracking where possible

JordyDeGroeve commented on src/06-time-tracking/time-tracking.apib:
```diff
@@ -0,0 +1,268 @@
+            + ids: `404c1777-6b61-4368-a388-0301dde8e72a`,`409e3446-96f2-4d9a-b20b-7f7cd055053e` (array[string], optional)
+            + user_id: `87982c96-f2fe-4b05-838c-ff42c0525758` (string, optional)
+            + work_type_id: `ce1287f0-0480-458d-b371-22862d864074` (string, optional)
+            + tracked_after: `2017-04-26T10:01:49+00:00` (string, optional)
```
Q: We filter on `started_after` in the code. Should we update this?

https://github.com/teamleadercrm/service-time-tracking/pull/60

WouterSioen replied:
Yep, you're right. That's what the Jira ticket said and I didn't double check with the documentation.

I think `started_after` is a bit more explicit though.

WouterSioen committed:
Rename timers.resume to timeTracking.resume

You resume a time tracking, which results in a timer, so this is the
more logical naming.

dietrichm commented on src/06-time-tracking/time-tracking.apib:
```diff
@@ -176,6 +176,23 @@ Delete tracked time.
 
 + Response 204
 
+### timeTracking.resume [POST /timers.resume]
```
B: should be `POST /timeTracking.resume` I think?

WouterSioen replied:
true that

 :ok: by @WouterSioen on 17/04/2018 16:41:40

mansilaxel commented on apiary.apib:
```diff
@@ -2192,6 +2192,277 @@ Update a milestone.
+
+    + Attributes (object)
+        + data (object)
+            + type: `timeTracking` (string)
```
Q: type: `timer` (string)? It is like this in `timers.start`

WouterSioen replied:
Nope, when you stop a timer, it get's saved, which makes it become a timeTracking

 :ok: by @WouterSioen on 17/04/2018 16:42:09

WouterSioen committed:
Fix url for timeTracking.resume endpoint

dietrichm committed:
Retrigger Concourse.

dietrichm committed:
Merge remote-tracking branch 'origin/master' into timetracking-endpoints

dietrichm commented on src/06-time-tracking/time-tracking.apib:
```diff
@@ -0,0 +1,268 @@
+            + type: `timer` (string)
+            + id: `f01adf4a-bb9b-45de-b231-615cd0e941de` (string)
+
+### timers.current [GET /timers.current]
```
PP: A header for "Timers" should be added before this line.

 :ok: by @dietrichm on 25/04/2018 17:19:10

dietrichm committed:
Add 'Timers' subsection.

Commit:
tracked_after / tracked_before

We changed these in implementation but this was not reflected in the api docs.

Commit:
replace tracked_* by started_*

implementation checks on started_after and _before, not tracked_after and _before

marijnvandevoorde commented on src/06-time-tracking/time-tracking.apib:
```diff
@@ -0,0 +1,270 @@
+                + ended_at: `2017-04-26T11:01:49+00:00` (string, required)
+            + Properties
+                + started_on: `2017-04-26` (string, required)
+                + duration: 3600 (number, required) - In seconds
```
B: This has to be removed from docs for now!!!

marijnvandevoorde commented on src/06-time-tracking/time-tracking.apib:
```diff
@@ -0,0 +1,270 @@
+        + One Of
+            + Properties
+                + started_at: `2017-04-26T10:01:49+00:00` (string, required)
+                + ended_at: `2017-04-26T11:01:49+00:00` (string, required)
```
B: not in there yet, but can be delivered quite fast I think, so let's keep this in here. But blocked until implemented

marijnvandevoorde commented on src/06-time-tracking/time-tracking.apib:
```diff
@@ -0,0 +1,270 @@
+                + started_at: `2017-04-26T10:01:49+00:00` (string, required)
+                + ended_at: `2017-04-26T11:01:49+00:00` (string, required)
+            + Properties
+                + started_on: `2017-04-26` (string, required)
```
B: remove this

marijnvandevoorde commented on src/06-time-tracking/time-tracking.apib:
```diff
@@ -0,0 +1,270 @@
+        + work_type_id: `a12f4289-3580-4780-bb6d-b4e9f84f80ad` (string, optional, nullable)
+        + One Of
+            + Properties
+                + started_at: `2017-04-26T10:01:49+00:00` (string, required)
```
B: until implemented. But let's keep this for now.

marijnvandevoorde reviewed (changes requested):
@jenssegers Can we remove the 2nd way of adding tracking (fuzzy, no started_at, only started_on). We don't have it yet.

dietrichm committed:
Merge remote-tracking branch 'origin/master' into timetracking-endpoints

JordyDeGroeve commented:
Can we remove the `.delete` endpoint until the side effects are handled?
If we want to release this asap.

WouterSioen committed:
Merge branch 'master' into timetracking-endpoints

WouterSioen committed:
Remove not implemented filters

WouterSioen committed:
Only allow started_at and duration for time tracking

WouterSioen committed:
Don't release timeTracking.delete for now

WouterSioen committed:
Build changes into the apiary file

WouterSioen commented:
Updated to be compatible with our current implementation, should be good to merge now.

JordyDeGroeve reviewed (approved):
👌

mansilaxel commented:
Not sure about using `tracked time` though.

marijnvandevoorde commented on src/06-time-tracking/time-tracking.apib:
```diff
@@ -100,12 +100,6 @@ Add tracked time.
     + Attributes (object)
         + work_type_id: `2175597d-484e-4a1c-a781-cbc3d9f893ba` (string, optional)
         + One Of
-            + Properties
```
This makes no sense, right? let's remove this level and require started_at & duration for now

marijnvandevoorde replied:
agreed to do this if needed, or add the ended_at soon so this is no longer an issue :-)
